subsys/nfc/t4t/isodep.c

```diff
@@ -0,0 +1,877 @@
+#define T4T_RATS_FSDI BIT(3)
+#define T4T_RATS_CMD_LEN 0x02
+
+#define T4T_FWT_ACTIVATION 71680
+#define T4T_FWT_DELTA 49152
+
+#define NFCA_T4T_FWT_T_FC (164 * 1356UL)
+#define T4T_FC_IN_MS 13560UL
+
+#define T4T_FWT_TO_MS(_fwt)                   \
+	ceiling_fraction((_fwt), T4T_FC_IN_MS)
+
+#define T4T_RATS_FDT (T4T_FWT_ACTIVATION +        \
+		T4T_FWT_DELTA + NFCA_T4T_FWT_T_FC)
+
+#define T4T_DID_MAX 14
+
+#define T4T_ATS_MAX_LEN 20
+#define T4T_ATS_T0_TA_PRESENT BIT(4)
+#define T4T_ATS_T0_TB_PRESENT BIT(5)
+#define T4T_ATS_T0_TC_PRESENT BIT(6)
+#define T4T_ATS_TA_EQUAL_DIVISOR BIT(7)
+#define T4T_ATS_T0_FSCI_MASK 0x0F
+#define T4T_ATS_TA_DIVISOR_PL 0x07
+#define T4T_ATS_TA_DIVISOR_PL_OFFSET 1
+#define T4T_ATS_TA_DIVISOR_LP_OFFSET 3
+#define T4T_ATS_TA_DIVISOR_LP 0x70
+#define T4T_ATS_TB_FWI_DEFAULT 0x04
+#define T4T_ATS_TB_FWI_MASK 0xF0
+#define T4T_ATS_TB_FWI_OFFSET 4
+#define T4T_ATS_TB_SFGI_MASK 0x0F
+#define T4T_ATS_TC_DID BIT(1)
+#define T4T_ATS_TC_NAD BIT(0)
+#define T4T_ATS_SFGT_DEFAULT 6780
+
+/* Convert FWI to FWT. NFC Forum Digital Specification 2.0 14.8.1. */
+#define T4T_FWI_TO_FWT(_fwi) (256u * 16u * (1 << (_fwi)))
+
+/* Calculate SFGT margin. NFC Forum Digital Specification 2.0 14.8.2. */
+#define T4T_SFGT_DELTA(_sfgi) (384 * (1 << (_sfgi)))
+#define ISODEP_BLOC_NUM_TOOGLE(block) ((block) ^ 0x01)
+
+enum isodep_frame {
+	ISODEP_FRAME_NONE,
+	ISODEP_FRAME_RATS,
+	ISODEP_FRAME_DESELECT,
+	ISODEP_FRAME_WTX_RESPONSE,
+	ISODEP_FRAME_I,
+	ISODEP_FRAME_I_CHAINING
+};
+
+enum isodep_state {
+	ISODEP_STATE_UNINITIALIZED    = 0,
+	ISODEP_STATE_INITIALIZED,
+	ISODEP_STATE_SELECTED,
+	ISODEP_STATE_TRANSFER
+};
+
+struct nfc_t4t_buf {
+	u8_t *data;
+	size_t len;
+	size_t buf_size;
+};
+
+struct nfc_t4t_err {
+	enum isodep_frame last_frame;
+	u8_t frame_retry_cnt;
+	u8_t wtx;
+};
+
+struct nfc_t4t_isodep {
+	atomic_t state;
+	struct nfc_t4t_isodep_tag tag;
+	struct nfc_t4t_buf tx_data;
+	struct nfc_t4t_buf rx_data;
+	struct nfc_t4t_err err_status;
+	u16_t fsd;
+	u8_t block_num;
+	u8_t retransmit_cnt;
+	const u8_t *transmit_data;
+	size_t transmit_len;
+	size_t transmitted_len;
+	bool chaining;
+	bool equal_divisor;
+	bool ats_expected;
+	bool first_transfer;
+};
+
+/* Map FSD value in terms of FSDI according to NFC Forum Digital Specification 2.0 14.16.1 */
+static const u16_t fsd_value_map[] = {16, 24, 32, 40, 48, 64, 96, 128, 256};
+
+static struct nfc_t4t_isodep t4t_isodep;
+static const struct nfc_t4t_isodep_cb *t4t_isodep_cb;
+static struct k_delayed_work isodep_work;
+static s64_t ats_received_time;
+
+static void isodep_transmission_clear(void)
+{
+	t4t_isodep.rx_data.len                = 0;
+	t4t_isodep.transmitted_len            = 0;
+	t4t_isodep.transmit_len               = 0;
+	t4t_isodep.chaining                   = false;
+	t4t_isodep.retransmit_cnt             = 0;
+	t4t_isodep.err_status.frame_retry_cnt = 0;
+	t4t_isodep.err_status.last_frame      = ISODEP_FRAME_NONE;
+}
+
+static bool did_include(u8_t *data)
```

Comment:
```
static size_t include_did(u8_t *data, size_t pos)
{
   data[pos] = 0xA;
   pos++
   if (IS_ENABLED(ADDME)) {
      data[pos] = 0xD;
      pos++
   }
   return pos;
}
```

Reply:
Currently you inform that element was added but information about number of added elements must be kept in multiple places.
It is less error pron to keep it in one place. See above case when config changes number of added elements.

Also `include_did` has better wording.